Sources/Rendering/OpenGL/Texture/index.js

```diff
@@ -1222,19 +1222,103 @@ function vtkOpenGLTexture(publicAPI, model) {
         halfFloatExt && model.openGLDataType === halfFloatExt.HALF_FLOAT_OES;
     }
 
-    if (!useHalfFloat) {
-      return false;
+    if (
+      useHalfFloat &&
+      (hasExactHalfFloat(offset, scale) || preferSizeOverAccuracy)
+    ) {
+      model.useHalfFloat = true;
+    } else {
+      model.useHalfFloat = false;
     }
+  }
 
-    // Don't consider halfFloat and convert back to Float when the range of data does not generate an accurate halfFloat
-    // AND it is not preferable to have a smaller texture than an exact texture.
-    if (!hasExactHalfFloat(offset, scale) && !preferSizeOverAccuracy) {
-      return false;
+  function processDataArray(dataArray, preferSizeOverAccuracy) {
+    const numComps = dataArray.getNumberOfComponents();
+    const dataType = dataArray.getDataType();
+    const data = dataArray.getData();
+
+    // Compute min max from array
+    // Using the vtkDataArray.getRange() enables caching
+    const minArray = new Array(numComps);
+    const maxArray = new Array(numComps);
+    for (let c = 0; c < numComps; ++c) {
+      const [min, max] = dataArray.getRange(c);
+      minArray[c] = min;
+      maxArray[c] = max;
     }
 
-    return true;
+    // Create a copy of scale and offset to avoid aliasing issues
+    // Original is read only, copy is read/write
+    // Use the copy as volumeInfo.scale and volumeInfo.offset
+    const scaleOffsets = computeScaleOffsets(minArray, maxArray, numComps);
+
+    const offset = [];
+    const scale = [];
+    for (let c = 0; c < numComps; ++c) {
+      offset[c] = 0.0;
+      scale[c] = 1.0;
+    }
+
+    setUseHalfFloat(
+      dataType,
+      scaleOffsets.offset,
+      scaleOffsets.scale,
+      preferSizeOverAccuracy
+    );
+
+    if (!model.useHalfFloat) {
+      publicAPI.getOpenGLDataType(dataType, true);
+    }
+
+    return {
+      numComps,
+      dataType,
+      data,
+      scaleOffsets: structuredClone(scaleOffsets),
```

Comment:
I do not think the `structuredClone` is called at the right time. It should be done AFTER setting scaleOffsets to the volumeInfo (in `create3DFilterableFromDataArray`)